### ISSUE TITLE
Too many redirections

### ISSUE THREAD

**Issue by zachantre** [nodogsplash/nodogsplash#356]
Hello,

I'm using OpenWRT 18.06 branch with NDS v3.2.1
I have tried to use FAS with nodogsplash, I have set port to 80, added the php files in /www/chrissi folder and fas_secure_enabled to 0.
With just this I get a too many redirections error. I can't understand why

here is my /etc/config/nodogsplash
https://pastebin.com/5quHATdb

here is the log with debug before connecting to hotspot
https://pastebin.com/fSqgebnf

and here is after i connect (where I get too many redirections error)
https://pastebin.com/MLKLj5DF

On the other side when using pre-auth with FAS it is working fine.
When I use BinAuth without FAS it is working fine.

Am I doing something wrong?

**Comment by bluewavenet** [comment]
Could you post your logs here instead of pastebin please. It is hard to follow there as there are so many advertisements. Also posting here will preserve the data for others to see later. Either paste or drag and drop a file (renamed as a .txt works well).

You say that Preauth works ok, so this indicates it is an issue with the FAS hosting which I assume is uhttpd.
Can you post your uhttpd config as well please.

It looks like the client does not have access to the FAS hosting.

When you connect a client device, does it automatically open a browser window?

**Comment by zachantre** [comment]
I just made a new compilation for 18.06 branch with NDS 3.3.1.
when I enable PreAuth with FAS it is working, I tried the example script with "Bill" and I can connect, browser opens new window for auth automatic.
I tried from lan to access FAS server and is working normal, of course it say "ERROR! Incomplete data passed from NDS"

In the previous logs I have noticed that when connecting to hotspot the redirect url is repeating the same string alot
`Sun Apr 14 21:52:04 2019 daemon.debug nodogsplash[3795]: splashpageurl: http://10.10.10.1:80/chrissi/fas.php?authaction=http://10.10.10.1:2050/nodogsplash_auth/?clientip=10.10.10.217&gatewayname=OpenWrt Nodogsplash&tok=1c10e4b9&redir=http%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir%3dhttp%3a%2f%2f10.10.10.1%2fchrissi%2ffas.php%3fredir`

here is the log of starting the NDS service and connecting a device
`Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: Reading and validating configuration file /tmp/etc/nodogsplash_cfg015847.conf
Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: Initializing signal handlers
Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: Detected gateway eth0.1010 at 10.10.10.1 (00:18:0a:42:bd:0f)
Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: Created web server on 10.10.10.1:2050
Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: Forwarding Authentication is Enabled.
Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: FAS URL is http://10.10.10.1:80/chrissi/fas.php
Sun Apr 14 23:57:06 2019 daemon.notice nodogsplash[2865]: Warning - Forwarding Authentication - Security is DISABLED.
Sun Apr 14 23:57:07 2019 daemon.notice nodogsplash[2865]: Initializing firewall rules
Sun Apr 14 23:57:20 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1) a4:50:46:ec:71:4a
Sun Apr 14 23:57:20 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1) 192.168.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:20 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1) a4:50:46:ec:71:4a
Sun Apr 14 23:57:20 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1) 192.168.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:20 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1) a4:50:46:ec:71:4a
Sun Apr 14 23:57:20 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1) 192.168.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1010) a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1010) a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1010) a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1010) a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPDISCOVER(eth0.1010) a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPOFFER(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPREQUEST(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.info dnsmasq-dhcp[2130]: DHCPACK(eth0.1010) 10.10.10.217 a4:50:46:ec:71:4a
Sun Apr 14 23:57:23 2019 daemon.notice nodogsplash[2865]: Adding 10.10.10.217 a4:50:46:ec:71:4a token bc1b0ef1 to client list
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]: could not encode url
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:24 2019 daemon.warn nodogsplash[2865]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:24 2019 kern.info kernel: [  429.033312] MHD-single[2868]: unhandled signal 11 at 00000000 nip b7a405b4 lr 1000a4f0 code 1
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Reading and validating configuration file /tmp/etc/nodogsplash_cfg015847.conf
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Initializing signal handlers
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Detected gateway eth0.1010 at 10.10.10.1 (00:18:0a:42:bd:0f)
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Created web server on 10.10.10.1:2050
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Forwarding Authentication is Enabled.
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: FAS URL is http://10.10.10.1:80/chrissi/fas.php
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Warning - Forwarding Authentication - Security is DISABLED.
Sun Apr 14 23:57:29 2019 daemon.notice nodogsplash[3018]: Initializing firewall rules
Sun Apr 14 23:57:30 2019 daemon.notice nodogsplash[3018]: Adding 10.10.10.217 a4:50:46:ec:71:4a token 1622ac18 to client list
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:30 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:34 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:37 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:44 2019 daemon.warn nodogsplash[3018]: could not encode url
Sun Apr 14 23:57:44 2019 daemon.warn nodogsplash[3018]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:44 2019 daemon.warn nodogsplash[3018]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:44 2019 daemon.warn nodogsplash[3018]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:44 2019 daemon.warn nodogsplash[3018]:  Query string exceeds the maximum of 2048 bytes so has been truncated.
Sun Apr 14 23:57:44 2019 kern.info kernel: [  449.597422] MHD-single[3019]: unhandled signal 11 at 00000000 nip b795a5b4 lr 1000a4f0 code 1
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: Reading and validating configuration file /tmp/etc/nodogsplash_cfg015847.conf
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: Initializing signal handlers
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: Detected gateway eth0.1010 at 10.10.10.1 (00:18:0a:42:bd:0f)
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: Created web server on 10.10.10.1:2050
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: Forwarding Authentication is Enabled.
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: FAS URL is http://10.10.10.1:80/chrissi/fas.php
Sun Apr 14 23:57:49 2019 daemon.notice nodogsplash[3214]: Warning - Forwarding Authentication - Security is DISABLED.
Sun Apr 14 23:57:50 2019 daemon.notice nodogsplash[3214]: Initializing firewall rules
Sun Apr 14 23:57:54 2019 daemon.notice nodogsplash[3214]: Adding 10.10.10.217 a4:50:46:ec:71:4a token 7ee7827b to client list
Sun Apr 14 23:57:54 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:54 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:54 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:54 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:54 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:54 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:55 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:55 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:55 2019 daemon.warn nodogsplash[3214]: could not encode url
Sun Apr 14 23:57:55 2019 daemon.warn nodogsplash[3214]: could not encode url
`

here is my uhttp config
`# Server configuration
config uhttpd main

	# HTTP listen addresses, multiple allowed
	list listen_http	0.0.0.0:80
	list listen_http	[::]:80

	# HTTPS listen addresses, multiple allowed
	list listen_https	0.0.0.0:443
	list listen_https	[::]:443

	# Redirect HTTP requests to HTTPS if possible
	option redirect_https	0

	# Server document root
	option home		/www

	# Reject requests from RFC1918 IP addresses
	# directed to the servers public IP(s).
	# This is a DNS rebinding countermeasure.
	option rfc1918_filter 1

	# Maximum number of concurrent requests.
	# If this number is exceeded, further requests are
	# queued until the number of running requests drops
	# below the limit again.
	option max_requests 3

	# Maximum number of concurrent connections.
	# If this number is exceeded, further TCP connection
	# attempts are queued until the number of active
	# connections drops below the limit again.
	option max_connections 100

	# Certificate and private key for HTTPS.
	# If no listen_https addresses are given,
	# the key options are ignored.
	option cert		/etc/uhttpd.crt
	option key		/etc/uhttpd.key

	# CGI url prefix, will be searched in docroot.
	# Default is /cgi-bin
	option cgi_prefix	/cgi-bin

	# List of extension->interpreter mappings.
	# Files with an associated interpreter can
	# be called outside of the CGI prefix and do
	# not need to be executable.
	list interpreter	".php=/usr/bin/php-cgi"
#	list interpreter	".cgi=/usr/bin/perl"

	# Lua url prefix and handler script.
	# Lua support is disabled if no prefix given.
	option lua_prefix	/luci
	option lua_handler	/usr/lib/lua/luci/sgi/uhttpd.lua

	# Specify the ubus-rpc prefix and socket path.
#	option ubus_prefix	/ubus
#	option ubus_socket	/var/run/ubus.sock

	# CGI/Lua timeout, if the called script does not
	# write data within the given amount of seconds,
	# the server will terminate the request with
	# 504 Gateway Timeout response.
	option script_timeout	60

	# Network timeout, if the current connection is
	# blocked for the specified amount of seconds,
	# the server will terminate the associated
	# request process.
	option network_timeout	30

	# HTTP Keep-Alive, specifies the timeout for persistent
	# HTTP/1.1 connections. Setting this to 0 will disable
	# persistent HTTP connections.
	option http_keepalive	20

	# TCP Keep-Alive, send periodic keep-alive probes
	# over established connections to detect dead peers.
	# The value is given in seconds to specify the
	# interval between subsequent probes.
	# Setting this to 0 will disable TCP keep-alive.
	option tcp_keepalive	1

	# Basic auth realm, defaults to local hostname
#	option realm	OpenWrt

	# Configuration file in busybox httpd format
#	option config	/etc/httpd.conf

	# Do not follow symlinks that point outside of the
	# home directory.
#	option no_symlinks	0

	# Do not produce directory listings but send 403
	# instead if a client requests an url pointing to
	# a directory without any index file.
#	option no_dirlists	0

	# Do not authenticate any ubus-rpc requests against
	# the ubus session/access procedure.
	# This is dangerous and should be always left off
	# except for development and debug purposes!
#	option no_ubusauth	0

	# For this instance of uhttpd use the listed httpauth
	# sections to require Basic auth to the specified
	# resources.
#	list httpauth prefix_user


# Defaults for automatic certificate and key generation
config cert defaults

	# Validity time
	option days		730

	# RSA key size
	option bits		2048

	# Location
	option country		ZZ
	option state		Somewhere
	option location		Unknown

	# Common name
	option commonname	'OpenWrt'

# config httpauth prefix_user
#	option prefix /protected/url/path
#	option username user
#	option password 'plaintext_or_md5_or_$p$user_for_system_user'
`

here is my nodogsplash config
`# The options available here are an adaptation of the settings used in nodogsplash.conf.
# See https://github.com/nodogsplash/nodogsplash/blob/master/resources/nodogsplash.conf

config nodogsplash
  # Set to 0 to disable nodogsplash
  option enabled 1

  # Set to 0 to disable hook that makes nodogsplash restart when the firewall restarts.
  # This hook is needed as a restart of Firewall overwrites nodogsplash iptables entries.
  option fwhook_enabled '1'

  # Serve the file splash.html from this directory
  option webroot '/etc/nodogsplash/htdocs'

  # Use plain configuration file
  #option config '/etc/nodogsplash/nodogsplash.conf'

  # Use this option to set the device nodogsplash will bind to.
  # The value may be an interface section in /etc/config/network or a device name such as br-lan.
  option gatewayinterface 'eth0.1010'

  # GatewayPort
  # Default: 2050
  #
  # Nodogsplash's own http server uses gateway address as its IP address.
  # The port it listens to at that IP can be set here; default is 2050.
  #
  option gatewayport '2050'


  option gatewayname 'OpenWrt Nodogsplash'
  option maxclients '250'

  # Enables debug output (0-7)
  #option debuglevel '7'

  # Client timeouts in minutes
  option preauthidletimeout '30'
  option authidletimeout '120'
  # Session Timeout is the interval after which clients are forced out (a value of 0 means never)
  option sessiontimeout '1200'

  # The interval in seconds at which nodogsplash checks client timeout status
  option checkinterval '600'

  # Enable BinAuth Support.
  # If set, a program is called with several parameters on authentication (request) and deauthentication.
  # Request for authentication:
  # $<BinAuth> auth_client <client_mac> '<username>' '<password>'
  #
  # The username and password values may be empty strings and are URL encoded.
  # The program is expected to output the number of seconds the client
  # is to be authenticated. Zero or negative seconds will cause the authentification request
  # to be rejected. The same goes for an exit code that is not 0.
  # The output may contain a user specific download and upload limit in KBit/s:
  # <seconds> <upload> <download>
  #
  # Called on authentication or deauthentication:
  # $<BinAuth> <*auth|*deauth> <incoming_bytes> <outgoing_bytes> <session_start> <session_end>
  #
  # "client_auth": Client authenticated via this script.
  # "client_deauth": Client deauthenticated by the client via splash page.
  # "idle_deauth": Client was deauthenticated because of inactivity.
  # "timeout_deauth": Client was deauthenticated because the session timed out.
  # "ndsctl_auth": Client was authenticated manually by the ndsctl tool.
  # "ndsctl_deauth": Client was deauthenticated by the ndsctl tool.
  # "shutdown_deauth": Client was deauthenticated by Nodogsplash terminating.
  #
  # Values session_start and session_start are in seconds since 1970 or 0 for unknown/unlimited.
  #
  #option binauth '/bin/myauth.sh'

  # Enable Forwarding Authentication Service (FAS)
  # If set redirection is changed from splash.html to a FAS (provided by the system administrator)
  # The value is the IP port number of the FAS
  option fasport '80'

  # Option: fasremoteip
  # Default: GatewayAddress (the IP of NDS)
  # If set, this is the remote ip address of the FAS.
  #option fasremoteip '46.32.240.41'

  # Option: faspath
  # Default: /
  # This is the path from the FAS Web Root to the FAS login page
  # (not the file system root).
  option faspath '/chrissi/fas.php'

  # Option: fas_secure_enabled
  # Default: 1
  # If set to "1", authaction and the client token are not revealed and it is the responsibility
  # of the FAS to request the token from NDSCTL.
  # If set to "0", the client token is sent to the FAS in clear text in the query string of the
  # redirect along with authaction and redir.
  option fas_secure_enabled '0'

  # Enable PreAuth Support.
  #
  # A simple login script is provided in the package.
  # This generates a login page asking for usename and email address.
  # User logins are recorded in the log file /tmp/ndslog.log
  # Details of how the script works are contained in comments in the script itself.
  #
  # PreAuth support allows FAS to call a local program or script with html served by NDS
  # If set, a program/script is called by the NDS FAS handler when:
  # 1. fasremopteip is not set,
  # 2. fasport is set to the gateway port
  # 3. faspath is set to /nodogsplash_preauth/
  #
  # Initially FAS appends its query string to faspath.
  #
  # The Preauth program will output html code that will be served to the client by NDS
  # Using html GET the Preauth program may call:
  # /nodogsplash_preauth/ to ask the client for more information
  # or
  # /nodogsplash_auth/ to authenticate the client
  #
  # The Preauth program should append at least the client ip to the query string
  # (using html input type hidden) for all calls to /nodogsplash_preauth/
  # It must also obtain the client token using ndsctl (or the original query string if fas_secure_enabled=0)
  # for NDS authentication when calling /nodogsplash_auth/
  #

  # Enable username/emailaddress login.
  # Note: fasport must be set to the same value as gatewayport (default = 2050)
  # Enable by uncommenting the following three lines
  #option fasport '2050'
  #option faspath '/nodogsplash_preauth/'
  #option preauth '/usr/lib/nodogsplash/login.sh'

  # Your router may have several interfaces, and you
  # probably want to keep them private from the gatewayinterface.
  # If so, you should block the entire subnets on those interfaces, e.g.:
  #list authenticated_users 'block to 192.168.0.0/16'
  #list authenticated_users 'block to 10.0.0.0/8'

  # Typical ports you will probably want to open up.
  #list authenticated_users 'allow tcp port 22'
  #list authenticated_users 'allow tcp port 53'
  #list authenticated_users 'allow udp port 53'
  #list authenticated_users 'allow tcp port 80'
  #list authenticated_users 'allow tcp port 443'
  # Or for happy customers allow all
  list authenticated_users 'allow all'

  # For preauthenticated users to resolve IP addresses in their
  # initial request not using the router itself as a DNS server,
  # Leave commented to help prevent DNS tunnelling
  #list preauthenticated_users 'allow tcp port 53'
  #list preauthenticated_users 'allow udp port 53'

  # Allow ports for SSH/Telnet/DNS/DHCP/HTTP/HTTPS
  list users_to_router 'allow tcp port 22'
  list users_to_router 'allow tcp port 23'
  list users_to_router 'allow tcp port 53'
  list users_to_router 'allow udp port 53'
  list users_to_router 'allow udp port 67'
  list users_to_router 'allow tcp port 80'

  # MAC addresses that are / are not allowed to access the splash page
  # Value is either 'allow' or 'block'. The allowedmac or blockedmac list is used.
  #option macmechanism 'allow'
  #list allowedmac '00:00:C0:01:D0:0D'
  #list allowedmac '00:00:C0:01:D0:1D'
  #list blockedmac '00:00:C0:01:D0:2D'

  # MAC addresses that do not need to authenticate
  #list trustedmac '00:00:C0:01:D0:1D'

  # Nodogsplash uses specific HEXADECIMAL values to mark packets used by iptables as a bitwise mask.
  # This mask can conflict with the requirements of other packages such as mwan3, sqm etc
  # Any values set here are interpreted as in hex format.
  #
  # List: fw_mark_authenticated
  # Default: 30000 (0011|0000|0000|0000|0000 binary)
  #
  # List: fw_mark_trusted
  # Default: 20000 (0010|0000|0000|0000|0000 binary)
  #
  # List: fw_mark_blocked
  # Default: 10000 (0001|0000|0000|0000|0000 binary)
  #
  #list fw_mark_authenticated '30000'
  #list fw_mark_trusted '20000'
  #list fw_mark_blocked '10000'
`
When

**Comment by bluewavenet** [comment]
Are you saying that on v3.3.1, you have it working with FAS and the demo phop script but on v3.2.1 it does not work?

Note : It would be much clearer if you posted the logs as a file. :-D

**Comment by bluewavenet** [comment]
Try changing the uhttpd port to 2080 and the fasport to match.

**Comment by zachantre** [comment]
In both version I have same behavior,

PreAuth + FAS = Working
only FAS = not working

I will make some more testing this afternoon, clean up the logs and post back.

I will also try on another device as well.

**Comment by bluewavenet** [comment]
I have it working in many installations so it must be something in your configuration, as I have not done anything particularly different although I am not using the demo script of course.
I will flash a basic 18.06 version and see if I can replicate this.

**Comment by zachantre** [comment]
Is it possible this happens because of kernel version? My router is Meraki MX60 and using 4.14 kernel.

**Comment by bluewavenet** [comment]
No not the kernel version. It is a bug in the demo fas.php somewhere. 
I can replicate it now!
I'll have a close look. .

**Comment by zachantre** [comment]
I see, ok.
Then maybe you can help me with what I need to do.
I would like to have an Android tablet which will provide a voucher code by pressing a button and then clients to use that code on their mobile to connect.
I guess this can happen with FAS and a custom php script that will create the code?

Which is the proper way to do that?

**Comment by bluewavenet** [comment]
Many ways to do that. 
eg the tab generates the voucher code and writes it into a fas table.
Fas then checks the if the voucher code is valid just as it does now for username/password (or would do if it did not have a bug... lol)

**Comment by bluewavenet** [comment]
It has been my intention to translate the default preauth script to php as an update to the current php demo script.
Maybe I should do this now ;-)

**Comment by bluewavenet** [comment]
I found the problem.
Changes to NDS firewall settings for other reasons (v3.2.1 onwards I think) have resulted in fasport = 80 being invalid when gatewayip is the same as the FAS server IP. I will look to see if this can be fixed, but I think it would be better to trap fasport=80 and give an error.

Anyway to get going make the following config changes

1. Add to /etc/config/uhttpd, the line:
`	list listen_http	0.0.0.0:2080`

2. Edit /etc/config/nodogsplash, updating fasport:
`  option fasport '2080'`

You must then restart uhttpd and nodogsplash:
```
service uhttpd restart
service nodogsplash restart
```

**Comment by zachantre** [comment]
At last it works :)
Thank you very much for your help...
Now I can try to make what I need...
In case I need help with the script can I post here?

**Comment by bluewavenet** [comment]
@zachantre 
>  In case I need help with the script can I post here?

I'll be happy to help, but it would be better to start a new issue at that time.
